### PR TITLE
Fit deaths by age

### PR DESCRIPTION
**Pull request by pabloperguz**
This PR implements fitting to deaths by age in Lancelot model. These are aggregated for age bands 1 to 10 (under 50s) and then broken down by 5-year age bands thereafter.

Assumptions are made on the distribution of hospital deaths for age bands CHW [18] and CHR [19]

Deaths in hospital for CHW [18] are assumed to be distributed as follows:
- 3/8 in the under 50s
- 1/8 in 50_54
- 2/8 in 55_59
- 2/8 in 60_64

Deaths in hospital for CHR [19] are assumed to be distributed as follows:
- 1/8 for each of age bands 65_69, 70_74 and 75_79
- 5/8 in the 80_plus

### CODE REVIEW

File: inst/odin/lancelot.R [hunk 1]
```diff
@@ -1665,10 +1665,43 @@ update(general_tot) <- new_general_tot
...
+initial(D_hosp_tot_70_74) <- 0
+delta_D_hosp_tot_70_74 <- delta_D_hosp[15] + delta_D_hosp[19] * 1 / 8
+update(D_hosp_tot_70_74) <- D_hosp_tot_70_74 + delta_D_hosp_tot_70_74
+
+initial(D_hosp_tot_75_79) <- 0
+delta_D_hosp_tot_75_79 <- delta_D_hosp[16] + delta_D_hosp[19] * 1 / 8
+update(D_hosp_tot_75_79) <- D_hosp_tot_75_79 + delta_D_hosp_tot_75_79
+
+initial(D_hosp_tot_80_plus) <- 0
+delta_D_hosp_tot_80_plus <- delta_D_hosp[17] + delta_D_hosp[19] * 5 / 8
```
Comment (edknock) [inline inst/odin/lancelot.R:1702]: Note that while you have split care home deaths (1/8, 1/8, 1/8, 5/8), CHR are drawn (0.05, 0.05, 0.15, 0.75).

Comment (edknock) [reply inst/odin/lancelot.R:1702]: (And if anything, ages of CHR deaths are likely to skew higher than age of CHR population)

File: R/lancelot.R [hunk 2]
```diff
@@ -1075,6 +1126,14 @@ lancelot_compare <- function(state, observed, pars) {
...
   ll_deaths <- ll_nbinom(observed$deaths,
                          pars$phi_death_hosp * model_deaths_hosp +
+                           pars$phi_death_hosp * model_deaths_hosp_0_49 +
+                           pars$phi_death_hosp * model_deaths_hosp_50_54 +
+                           pars$phi_death_hosp * model_deaths_hosp_55_59 +
+                           pars$phi_death_hosp * model_deaths_hosp_60_64 +
+                           pars$phi_death_hosp * model_deaths_hosp_65_69 +
+                           pars$phi_death_hosp * model_deaths_hosp_70_74 +
+                           pars$phi_death_hosp * model_deaths_hosp_75_79 +
+                           pars$phi_death_hosp * model_deaths_hosp_80_plus +
```
Comment (edknock) [inline R/lancelot.R:1136]: Here we should have just `pars$phi_death_hosp * model_deaths_hosp` or just 
```pars$phi_death_hosp * model_deaths_hosp_0_49 +
 pars$phi_death_hosp * model_deaths_hosp_50_54 +
 pars$phi_death_hosp * model_deaths_hosp_55_59 +
 pars$phi_death_hosp * model_deaths_hosp_60_64 +
 pars$phi_death_hosp * model_deaths_hosp_65_69 +
 pars$phi_death_hosp * model_deaths_hosp_70_74 +
 pars$phi_death_hosp * model_deaths_hosp_75_79 +
 pars$phi_death_hosp * model_deaths_hosp_80_plus

Comment (edknock) [reply R/lancelot.R:1136]: Also none of the age disaggregated hospital death streams are contributing to the likelihood currently beyond this

Comment (pabloperguz) [reply R/lancelot.R:1136]: Ah, sorry about this. Done

File: R/lancelot.R [hunk 3]
```diff
@@ -740,6 +740,14 @@ lancelot_index <- function(info) {
                   deaths_carehomes_inc = index[["D_carehomes_inc"]],
                   deaths_comm_inc = index[["D_comm_inc"]],
                   deaths_hosp_inc = index[["D_hosp_inc"]],
+                  deaths_hosp_inc_0_49 = index[["D_hosp_inc_0_49"]],
```
Comment (edknock) [inline R/lancelot.R:743]: I think we should change ordering to `deaths_hosp_0_49_inc` (and also `D_hosp_0_49_inc` in the odin code), it's consistent with how `_inc` always comes last (e.g. `sympt_cases_15_24_inc`)

File: inst/odin/lancelot.R [hunk 4]
```diff
@@ -1665,10 +1665,43 @@ update(general_tot) <- new_general_tot
 initial(hosp_tot) <- 0
 update(hosp_tot) <- new_ICU_tot + new_general_tot
 
+## cumulative deaths in hospital by age
 initial(D_hosp_tot) <- 0
 delta_D_hosp_tot <- sum(delta_D_hosp)
 update(D_hosp_tot) <- D_hosp_tot + delta_D_hosp_tot
 
+initial(D_hosp_tot_0_49) <- 0
```
Comment (edknock) [inline inst/odin/lancelot.R:1673]: Again probably order this `D_hosp_0_49_tot`

**Comment by edknock** [comment]
Beyond above comments, a couple of things:

1. Please update the compiled compare - some of the tests in test-lancelot.R are failing locally for me (not sure why they're not failing on Github).
2. Please put in checks to stop double fitting hospital deaths - no fitting to age aggregated hospital deaths if you fit to any age specific hospital deaths, and no fitting to all deaths aggregated if you fit to any age specific hospital deaths

File: tests/testthat/test-lancelot-support.R [hunk 5]
```diff
@@ -533,6 +536,14 @@ test_that("lancelot_compare combines likelihood correctly", {
...
     deaths_comm_inc = 1:6,
     deaths_hosp_inc = 3:8,
+    deaths_hosp_inc_0_49 = 3:8,
+    deaths_hosp_inc_50_54 = 3:8,
+    deaths_hosp_inc_55_59 = 3:8,
+    deaths_hosp_inc_60_64 = 3:8,
+    deaths_hosp_inc_65_69 = 3:8,
+    deaths_hosp_inc_70_74 = 3:8,
+    deaths_hosp_inc_75_79 = 3:8,
+    deaths_hosp_inc_80_plus = 3:8,
```
Comment (edknock) [inline tests/testthat/test-lancelot-support.R:546]: might want to have different values for each of the above

File: tests/testthat/test-lancelot-support.R [hunk 6]
```diff
@@ -554,6 +565,14 @@ test_that("lancelot_compare combines likelihood correctly", {
...
     deaths_carehomes = 4,
     deaths_hosp = 5,
+    deaths_hosp_0_49 = 5,
+    deaths_hosp_50_54 = 5,
+    deaths_hosp_55_59 = 5,
+    deaths_hosp_60_64 = 5,
+    deaths_hosp_65_69 = 5,
+    deaths_hosp_70_74 = 5,
+    deaths_hosp_75_79 = 5,
+    deaths_hosp_80_plus = 5,
```
Comment (edknock) [inline tests/testthat/test-lancelot-support.R:575]: again, might want to have different values for each of the above

File: inst/odin/lancelot.R (hunk 1 above)
Comment (pabloperguz) [reply inst/odin/lancelot.R]: Done

File: R/lancelot.R (hunk 3 above)
Comment (pabloperguz) [reply R/lancelot.R]: Done

File: inst/odin/lancelot.R (hunk 4 above)
Comment (pabloperguz) [reply inst/odin/lancelot.R]: Done

File: tests/testthat/test-lancelot-support.R (hunk 6 above)
Comment (pabloperguz) [reply tests/testthat/test-lancelot-support.R]: Done

File: tests/testthat/test-lancelot-support.R (hunk 5 above)
Comment (pabloperguz) [reply tests/testthat/test-lancelot-support.R]: Done

File: R/lancelot.R [hunk 7]
```diff
@@ -2163,6 +2236,10 @@ lancelot_prepare_data <- function(data) {
     }
   } else {
     lapply(split(data, data$population), function(x) {
+      if (check_deaths(x) > 1) {
+        stop(sprintf("cannot fit to all ages aggregated for deaths if fitting
```
Comment (richfitz) [inline R/lancelot.R:2240]: Can you please add a test for this condition

Comment (pabloperguz) [reply R/lancelot.R:2240]: Thanks Rich, will do

File: inst/odin/lancelot.R [hunk 8]
```diff
@@ -1665,10 +1665,43 @@ update(general_tot) <- new_general_tot
 initial(hosp_tot) <- 0
 update(hosp_tot) <- new_ICU_tot + new_general_tot
 
+## cumulative deaths in hospital by age
 initial(D_hosp_tot) <- 0
 delta_D_hosp_tot <- sum(delta_D_hosp)
 update(D_hosp_tot) <- D_hosp_tot + delta_D_hosp_tot
 
+initial(D_hosp_0_49_tot) <- 0
+delta_D_hosp_0_49_tot <- sum(delta_D_hosp[1:10]) + delta_D_hosp[18] * 3 / 8
```
Comment (edknock) [inline inst/odin/lancelot.R:1674]: It strikes me that the proportion of CHW deaths that we are assuming are for each group is a parameter that's better off not hardcoded. Same for the CHR parameters.

Comment (edknock) [reply inst/odin/lancelot.R:1674]: And also in hindsight, the ones for pillar 2 by age

File: tests/testthat/test-lancelot.R [hunk 9]
```diff
@@ -246,6 +254,10 @@ test_that("Test compiled lancelot components", {
     c(icu = 50),
     c(general = 50),
     c(hosp = 50),
+    c(deaths_hosp_0_49 = 1, deaths_hosp_50_54 = 1, deaths_hosp_55_59 = 1,
+      deaths_hosp_60_64 = 2, deaths_hosp_65_69 = 2, deaths_hosp_70_74 = 3,
+      deaths_hosp_75_79 = 5, deaths_hosp_80_plus = 10,
+      deaths_hosp = 30, deaths_carehomes = 10, deaths_comm = 10),
```
Comment (edknock) [inline tests/testthat/test-lancelot.R:260]: Can you please separate each death stream out on its own. Same for deaths_hosp, deaths_carehomes, deaths_comm and deaths_non_hosp below please - they all should have already been on their own. The only need to pair here is when data streams pair up for binomial/beta-binomial fitting

**Comment by edknock** [comment]
Please add to this test to check incidence calculation (it would be good to add those for cases by age too)
https://github.com/mrc-ide/sircovid/blob/master/tests/testthat/test-lancelot.R#L145

**Comment by edknock** [comment]
Please add a test like this to check hospital deaths by age sum up to the total https://github.com/mrc-ide/sircovid/blob/master/tests/testthat/test-lancelot-check.R#L883

File: tests/testthat/test-lancelot.R [hunk 10]
```diff
@@ -246,6 +270,17 @@ test_that("Test compiled lancelot components", {
...
+    c(deaths_hosp_50_54 = 1),
+    c(deaths_hosp_55_59 = 1),
+    c(deaths_hosp_60_64 = 2),
+    c(deaths_hosp_65_69 = 2),
+    c(deaths_hosp_70_74 = 3),
+    c(deaths_hosp_75_79 = 5),
+    c(deaths_hosp_80_plus = 10),
+    c(deaths_hosp = 30),
+    c(deaths_carehomes = 10),
+    c(deaths_comm = 10),
```
Comment (edknock) [inline tests/testthat/test-lancelot.R:283]:
```suggestion
    c(deaths_comm = 10),
    c(deaths_non_hosp = 20),
```

**Comment by edknock** [comment]
Please add here to ensure we don't fit to `deaths` if we are fitting to any age specific hospital deaths https://github.com/mrc-ide/sircovid/blob/6996af88c3e4cc2fee4aa7478f14ff91b69ea668/R/lancelot.R#L2167-L2172

**Pull request merged**